src/lldp/lldppkt.rs

```diff
@@ -0,0 +1,1086 @@
+#[derive(Debug)]
+pub struct LLDPPacket {
+    pub tlvs: Vec<LLDPTLV>,
+}
+
+impl LLDPPacket {
+    fn validate_format(&self) -> Result<(), std::io::Error> {
+        if self.tlvs.len() < 4 {
+            return Err(std::io::Error::new(
+                ErrorKind::InvalidData,
+                "At least 4 mandatory TLVs are required in an LLDP PDU.",
+            ));
+        }
+
+        // Safe to unwrap four elements since we checked above.
+        let first: &LLDPTLV = self.tlvs.get(0).unwrap();
+        let second: &LLDPTLV = self.tlvs.get(1).unwrap();
+        let third: &LLDPTLV = self.tlvs.get(2).unwrap();
+        let last: &LLDPTLV = self.tlvs.get(self.tlvs.len() - 1).unwrap();
+
+        if !matches!(first, LLDPTLV::ChassisID(_)) {
+            return Err(std::io::Error::new(
+                ErrorKind::InvalidData,
+                format!("Expected first TLV to be ChassisID but got {}", first),
+            ));
+        }
+        if !matches!(second, LLDPTLV::PortID(_)) {
+            return Err(std::io::Error::new(
+                ErrorKind::InvalidData,
+                format!("Expected second TLV to be PortID but got {}", second),
+            ));
+        }
+        if !matches!(third, LLDPTLV::TTL(_)) {
+            return Err(std::io::Error::new(
+                ErrorKind::InvalidData,
+                format!("Expected third TLV to be TTL but got {}", third),
+            ));
+        }
+        if !matches!(last, LLDPTLV::EndOfLLDPPDU()) {
+            return Err(std::io::Error::new(
+                ErrorKind::InvalidData,
+                format!("Expected last TLV to be EndOfLLDPPDU but got {}", last),
+            ));
+        }
+
+        Ok(())
+    }
+}
+
+impl ToWire for LLDPPacket {
+    fn to_wire(&self) -> Result<Vec<u8>, std::io::Error> {
+        self.validate_format()?;
+        let mut pdu = Vec::new();
+
+        for tlv in &self.tlvs {
+            pdu.append(&mut tlv.to_wire()?);
+        }
+
+        Ok(pdu)
+    }
+}
+
+impl FromWire for LLDPPacket {
+    fn from_wire(
+        buf: &mut pktparser::Buffer<'_>,
+    ) -> std::result::Result<Self, pktparser::ParseError> {
+        let mut tlvs = Vec::new();
+        while buf.remaining() > 0 {
+            tlvs.push(LLDPTLV::from_wire(buf)?);
+            if matches!(tlvs[tlvs.len() - 1], LLDPTLV::EndOfLLDPPDU()) {
+                return Ok(LLDPPacket { tlvs });
+            }
+        }
+        Err(pktparser::ParseError::InvalidArgument(
+            "Malformed LLDP packet: missing End of LLDP PDU TLV".to_string(),
+        ))
+    }
+}
+
+impl std::fmt::Display for LLDPPacket {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        write!(f, "LLDP [\n")?;
+        for tlv in &self.tlvs {
+            write!(f, "\t{}\n", tlv)?;
+        }
+        write!(f, "]")
+    }
+}
+
+#[derive(Debug)]
+pub struct LLDPTLVHeader {
+    pub r#type: u8,  // 7 bits on the wire.
+    pub length: u16, // 9 bits on the wire
+}
+
```

Comment:
you may want a:
```rust
impl LldpTlvHeader {
 pub fn new(ty: u8, len: usize) -> Self {
   Self { ty, len: (len +1) as u16 }
 }
}
```

which will tidy up the serialisers below.

Reply:
This seems mostly redundant now that the header logic is all centralized in one place.